Azure IPAM: Replace subscription-wide VNet enumeration with targeted subnet queries to eliminate NRP throttling

yuecong opened the pull request:
## Summary

This PR implements a targeted subnet discovery optimization for Azure IPAM that eliminates Azure NRP throttling by replacing subscription-wide VNet enumeration with targeted subnet queries.

### Key Changes

- **Three-phase subnet discovery strategy:**
  1. Query all node instances (existing method)
  2. Extract unique subnet IDs from node network interfaces
  3. Query only the specific subnets that nodes actually use


- **Performance improvements:**
  - Reduces API calls from O(n*m) to O(k) where k = unique subnets used by nodes
  - Eliminates subscription-wide VNet enumeration that causes throttling
  - Maintains backward compatibility with fallback to full VNet discovery

### Test Results

- All API tests pass including new parseSubnetID validation tests
- Azure IPAM functionality preserved with enhanced logging
- Existing IPAM allocation tests demonstrate continued functionality

### Breaking Changes

None - maintains full backward compatibility with existing IPAM behavior.

### Performance Impact

Significantly improved performance in large Azure environments with many VNets while maintaining the same IPAM functionality. The optimization shows "targeted_subnets" count in logs for visibility.

## Test plan

- [x] Unit tests for parseSubnetID function with 8 validation scenarios
- [x] Azure API tests pass
- [x] Mock implementation updated for GetNodesSubnets
- [x] IPAM allocation tests demonstrate preserved functionality

I deployed this change to the main branch to our internal clusters and confirmed that this works. Also for the 1.16 version, we have used it in production for more than 2 months in ~100 clusters

```release-note
Azure IPAM: Optimize subnet discovery to eliminate Azure NRP throttling by using targeted subnet queries instead of subscription-wide VNet enumeration, significantly improving performance in large Azure environments with many VNets.
```

tamilmani1989 commented on pkg/azure/ipam/instances.go:
```diff
@@ -87,64 +88,117 @@ func (m *InstancesManager) Resync(ctx context.Context) time.Time {
+	
+	instances.ForeachAddress("", func(instanceID, interfaceID, ip, poolID string, address ipamTypes.Address) error {
+		if poolID != "" {
+			subnetIDs[poolID] = struct{}{}
```
why this has to be map and not array of strings?

yuecong replied:
The map is used to automatically deduplicate subnet IDs since the same poolID can appear multiple times when
  iterating through addresses, avoiding O(n) lookups that would be needed with an array.

tamilmani1989 replied:
got it, can you add a comment in code on why we use map

tamilmani1989 replied:
one follow-up question: if there are multiple poolIds, which poolId will be used ?

yuecong replied:
for one instance, I think we can only have one poolID (which is subnet ID) in this case. Let me know if I misunderstood your question here

tamilmani1989 replied:
there can be multiple poolIDs if its multi nic VM

yuecong replied:
> got it, can you add a comment in code on why we use map

this is addressed

yuecong replied:
> there can be multiple poolIDs if its multi nic VM

Thanks for the clarification. my previous comment on `only have one poolID` is not correct.
I checked the code if it handles multiple NICs correctly and TL;DR is it does handle that correctly

  1. The extractSubnetIDs() function uses a map to collect all unique subnet IDs across all interfaces
  2. ForeachAddress() iterates through every interface of an instance (pkg/ipam/types/types.go:611)
  3. Each interface's addresses can have different subnet IDs (poolIDs)
  4. The map deduplicates while preserving all unique subnets


  Example Scenario

  VM with 2 NICs:
    NIC-1 in subnet-A → poolID = "subnet-a"
    NIC-2 in subnet-B → poolID = "subnet-b"

  extractSubnetIDs() correctly returns: ["subnet-a", "subnet-b"]
  GetNodesSubnets() queries both subnets

tamilmani1989 replied:
ok, how it will know which subnet to be used for Pod IPs?

yuecong replied:
> ok, how it will know which subnet to be used for Pod IPs?

as here is the native mode, so the IP allocation is from the Azure subnet, the behavior will be the subnet where the node is in to be used for the pods on that node

tamilmani1989 commented on pkg/azure/ipam/instances.go:
```diff
@@ -87,64 +88,117 @@ func (m *InstancesManager) Resync(ctx context.Context) time.Time {
+	subnetIDs := make(map[string]struct{})
+	
+	instances.ForeachAddress("", func(instanceID, interfaceID, ip, poolID string, address ipamTypes.Address) error {
+		if poolID != "" {
```
can you share offical azure doc where it says poolID is ARM resource id of subnet?

yuecong replied:
By the way, poolID has been the ARM resource ID before this PR. some reference
```
  1. pkg/azure/api/api.go:351-352 (existing code, not modified by PR):
  if ip.Properties.Subnet != nil {
      addr.Subnet = *ip.Properties.Subnet.ID
    This has always populated address.Subnet with the subnet's ARM resource ID from Azure API.
  2. pkg/azure/types/types.go:188 (existing code, not modified by PR):
  if err := fn(id, a.ID, address.IP, address.Subnet, address); err != nil {
     This has always passed address.Subnet as the poolID parameter.
  ```

  According to the Azure REST API documentation, ip.Properties.Subnet.ID contains the ARM resource ID
  in this format:
  /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/virtualNetworks/{vnetName}/subnets/{subnetName}

  Official docs: https://learn.microsoft.com/en-us/rest/api/virtualnetwork/network-interfaces/get

tamilmani1989 commented on pkg/azure/ipam/instances.go:
```diff
@@ -87,64 +88,117 @@ func (m *InstancesManager) Resync(ctx context.Context) time.Time {
+
+	// Phase 3: Re-parse instances with subnet details to populate CIDR and gateway info
+	if len(subnets) > 0 {
+		instances, err = m.api.GetInstances(ctx, subnets)
```
why GetInstances called again here?

yuecong replied:
The first call extracts the subnet IDs so you know which subnets to query. The second call enriches the data
  with CIDR and gateway information needed for IP allocation.
```
  First Call (Line 166) - Phase 1:

  instances, err := m.api.GetInstances(ctx, ipamTypes.SubnetMap{})
                                           ^^^^^^^^^^^^^^^^^
                                           EMPTY subnet map

  Second Call (Line 185) - Phase 3:

  instances, err = m.api.GetInstances(ctx, subnets)
                                           ^^^^^^^
                                           POPULATED subnet map
```

tamilmani1989 replied:
i feel we can avoid duplicate calls if we do `foreachinterface` and parse everything in one go.

yuecong replied:
I am not sure if it will work.
  1. GetInstances() already iterates over ALL interfaces via listAllNetworkInterfaces() (pkg/azure/api/api.go:384)
  2. Subnet data comes from a separate Azure API - there's no single API call that returns both interfaces AND subnet CIDR details
  3. Parsing happens during iteration, but subnet details aren't available yet
  
FYI, this is why we need to call GetInstances with different parameters. this is the way I can think of is optimal in terms of Azure API calls. A bit nuanced and here is the details.

  Phase 1 (Line 166): GetInstances(ctx, ipamTypes.SubnetMap{})
  - Fetches ALL network interfaces with an empty subnet map
  - Returns instances with subnet IDs but without CIDR/Gateway info
  - Used to discover which subnet IDs are actually in use

  Phase 2 (Line 176): GetNodesSubnets(ctx, subnetIDs)
  - Queries only the specific subnets discovered in Phase 1
  - Returns subnet details including CIDR blocks and available IPs

  Phase 3 (Line 185): GetInstances(ctx, subnets)
  - Re-fetches ALL network interfaces, but now with subnet details
  - This time populates CIDR and Gateway fields in the parsed interfaces

  Why Two Calls Are Needed

  Looking at parseInterface() in pkg/azure/api/api.go:353-361, the CIDR and Gateway are only populated when:
```
  if subnet, ok := subnets[addr.Subnet]; ok {
      if subnet.CIDR.IsValid() {
          i.CIDR = subnet.CIDR.String()  // ← Only set if subnet map has data
      }
      if gateway := deriveGatewayIP(subnet.CIDR.Addr()); gateway != "" {
          i.GatewayIP = gateway
          i.Gateway = gateway
      }
  }
```

tamilmani1989 replied:
got it, thanks. the 1st call `getInstances` retrieves all networkInterfaces in that subscription or retrieve only networkInterface associated with that node.

yuecong replied:
I am optimizing it to see if we can remove the 2nd call on getInstancs

yuecong replied:
this is addressed

tamilmani1989 replied:
> GetInstances() already iterates over ALL interfaces via listAllNetworkInterfaces()

Does it list of network interfaces in a subscription? if so, how this is better than listing all virtual networks?

yuecong replied:
> > GetInstances() already iterates over ALL interfaces via listAllNetworkInterfaces()
> 
> Does it list of network interfaces in a subscription? if so, how this is better than listing all virtual networks?

it does not list all the network interfaces for all resource groups in a subscription but only list the network interfaces in the resource group where the node are in.

tamilmani1989 commented on pkg/azure/api/api.go:
```diff
@@ -491,6 +512,122 @@ func (c *Client) GetVpcsAndSubnets(ctx context.Context) (ipamTypes.VirtualNetwor
+}
+
+// GetNodesSubnets retrieves subnet information for specific subnet IDs
+func (c *Client) GetNodesSubnets(ctx context.Context, subnetIDs []string) (ipamTypes.SubnetMap, error) {
```
can this api be `GetSubnetInfobySubnetID`?

yuecong replied:
Would you agree on naming it GetSubnetsByIDs? btw, I like a better naming. If you prefer the name of GetSubnetInfoBySubnetID, I can change it as well, which is better than my original naming. :)

tamilmani1989 replied:
fine with either

yuecong replied:
this is addressed

tamilmani1989 commented on pkg/azure/ipam/instances.go:
```diff
@@ -87,64 +88,117 @@ func (m *InstancesManager) Resync(ctx context.Context) time.Time {
+func (m *InstancesManager) extractSubnetIDs(instances *ipamTypes.InstanceMap) []string {
+	subnetIDs := make(map[string]struct{})
+	
+	instances.ForeachAddress("", func(instanceID, interfaceID, ip, poolID string, address ipamTypes.Address) error {
```
instead of using `ForeachAddress` can't we use `ForeachInterface` and it has cidr and gateway info already and no need to query getsubnet arm call.  for eg: https://github.com/cilium/cilium/blob/8e8421dd3ee70e2596d83afc29e245fbfa96ed30/pkg/azure/ipam/node.go#L49

yuecong replied:
I think this question is related to the above question.

  The Key Issue: CIDR and Gateway Are NOT Available After Phase 1

  Looking at how parseInterface() works (pkg/azure/api/api.go:353-361):
```
  if ip.Properties.Subnet != nil {
      addr.Subnet = *ip.Properties.Subnet.ID  // ← Always populated
      if subnet, ok := subnets[addr.Subnet]; ok {  // ← Only enters if subnets map is NOT empty
          if subnet.CIDR.IsValid() {
              i.CIDR = subnet.CIDR.String()  // ← Only set when subnets map provided
          }
          if gateway := deriveGatewayIP(subnet.CIDR.Addr()); gateway != "" {
              i.GatewayIP = gateway
              i.Gateway = gateway
          }
      }
  }
```
  After Phase 1 (line 166): GetInstances(ctx, ipamTypes.SubnetMap{})
  - Subnet IDs are populated (e.g., /subscriptions/.../subnets/subnet-1)
  - CIDR and Gateway fields are EMPTY because the subnets map is empty

  After Phase 3 (line 185): GetInstances(ctx, subnets)
  - Subnet IDs are populated
  - CIDR and Gateway are NOW populated because subnets map has data

  **Why ForeachInterface Won't Help**

  The example in node.go:48-52 that you referenced uses ForeachInterface after all 3 phases are complete, so at that point the interfaces DO have CIDR and Gateway populated. But that's not the case after Phase 1.

  The issue is that:
  1. Azure's network interface API returns subnet IDs only (not CIDR)
  2. To get subnet CIDR blocks, you must call the Subnets API separately
  3. GetNodesSubnets (Phase 2) is required to fetch those CIDR details

  Cannot Avoid the GetNodesSubnets Call

  No, we cannot use ForeachInterface to avoid the GetNodesSubnets ARM call because:

  - After Phase 1, interfaces have subnet IDs but no CIDR/Gateway
  - The Azure network interface API doesn't include subnet CIDR in its response
  - You must make a separate Subnets API call to get CIDR information
  - That's exactly what GetNodesSubnets does

  The three-phase approach is necessary given Azure's API design:
  Phase 1: Discover which subnet IDs are in use (from interface responses)
  Phase 2: Fetch CIDR details for those specific subnets (targeted query)
  Phase 3: Re-parse interfaces with subnet CIDR data to populate Gateway/CIDR fields


hope this helps explain the details here.

tamilmani1989 commented on pkg/azure/api/api.go:
```diff
@@ -19,6 +20,7 @@ import (
 	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
 	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/compute/armcompute/v7"
 	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/network/armnetwork/v7"
+	"github.com/sirupsen/logrus"
```
i think all cilium packages are now using slog. can we use the same?

yuecong replied:
sure. I can change this.

yuecong replied:
this is changed

tamilmani1989 commented on pkg/azure/api/api.go:
```diff
@@ -38,18 +40,30 @@ const (
 	interfacesListVirtualMachineScaleSetVMNetworkInterfaces = "Interfaces.ListVirtualMachineScaleSetVMNetworkInterfaces"
+
+	// Constants for subnet ID parsing
+	expectedCaptureGroups = 4
```
can it be `expectedSubnetResourceIDSplit` ?

yuecong replied:
Done. Renamed to expectedSubnetResourceIDSplit.

tamilmani1989 commented on pkg/azure/api/api.go:
```diff
@@ -507,6 +571,122 @@ func (c *Client) GetVpcsAndSubnets(ctx context.Context) (ipamTypes.VirtualNetwor
+	}
+
+	subnet := &result.Subnet
+	if subnet.ID == nil {
```
can subnet be nil?

yuecong replied:
No, subnet cannot be nil here. Since result.Subnet is an embedded struct (value type, not a pointer), &result.Subnet always returns a valid pointer. The subnet.ID == nil check is still valid defensive programming to handle potential edge cases in the API response.

tamilmani1989 reviewed (commented):
@yuecong overall lgtm. added minor comments. Can you fix all lint errors

tamilmani1989 commented on pkg/azure/api/api.go:
```diff
@@ -507,6 +571,122 @@ func (c *Client) GetVpcsAndSubnets(ctx context.Context) (ipamTypes.VirtualNetwor
 }
 
+// parseSubnetID extracts resource group, virtual network, and subnet names from an Azure subnet ID
+func parseSubnetID(subnetID string) (resourceGroupName, vnetName, subnetName string, err error) {
```
@yuecong should we use something like this to get resourceIds https://github.com/cilium/cilium/pull/43006

yuecong replied:
Good suggestion. Refactored parseSubnetID to use arm.ParseResourceID from the Azure SDK instead of custom regex. This is consistent with how we parse resource IDs elsewhere in this file (lines 434, 453, 842, etc.). 

Added validation for provider namespace (Microsoft.Network), resource type, and presence of resource group. All existing tests pass.

tamilmani1989 replied:
minor: thanks, can we specify expected format of subnet arm resource id in comment

yuecong committed:
Azure IPAM: Replace VNet enumeration with targeted subnet queries

This PR implements a targeted subnet discovery optimization for Azure IPAM that eliminates Azure NRP throttling by replacing subscription-wide VNet enumeration with targeted subnet queries.

Key Changes:

- Three-phase subnet discovery strategy:
  1. Query all node instances (existing method)
  2. Extract unique subnet IDs from node network interfaces
  3. Query only the specific subnets that nodes actually use

- Performance improvements:
  - Reduces API calls from O(n*m) to O(k) where k = unique subnets used by nodes
  - Eliminates subscription-wide VNet enumeration that causes throttling
  - Maintains backward compatibility with fallback to full VNet discovery

Test Results:

- All API tests pass including new parseSubnetID validation tests
- Azure IPAM functionality preserved with enhanced logging
- Existing IPAM allocation tests demonstrate continued functionality

Breaking Changes:

None - maintains full backward compatibility with existing IPAM behavior.

Performance Impact:

Significantly improved performance in large Azure environments with many VNets while maintaining the same IPAM functionality. The optimization shows "targeted_subnets" count in logs for visibility.

Test plan:

- Unit tests for parseSubnetID function with 8 validation scenarios
- Azure API tests pass
- Mock implementation updated for GetNodesSubnets
- IPAM allocation tests demonstrate preserved functionality

I deployed this change to the main branch to our internal clusters and confirmed that this works. Also for the 1.16 version, we have used it in production for more than 2 months in ~100 clusters

Signed-off-by: yuecong <cong@databricks.com>

pippolo84 commented on pkg/azure/api/api.go:
```diff
@@ -507,6 +564,147 @@ func (c *Client) GetVpcsAndSubnets(ctx context.Context) (ipamTypes.VirtualNetwor
+	// /subscriptions/.../virtualNetworks/{vnet}/subnets/{subnet} -> /subscriptions/.../virtualNetworks/{vnet}
+	if strings.Contains(subnetID, "/subnets/") {
+		return subnetID[:strings.LastIndex(subnetID, "/subnets/")]
+	}
```
Nit: no need for `strings.Contains` before `strings.LastIndex`:

```go
	idx := strings.LastIndex(subnetID, "/subnets/")
	if idx == -1 {
		return ""
	}
	return subnetID[:idx]
```

pippolo84 commented on pkg/azure/api/api.go:
```diff
@@ -408,13 +437,41 @@ func (c *Client) GetInstance(ctx context.Context, subnets ipamTypes.SubnetMap, i
+// ParseInterfacesIntoInstance parses network interfaces into an Instance
+// This allows re-parsing the same network interface data with different subnet maps
+// without making additional Azure API calls
+func (c *Client) ParseInterfacesIntoInstance(networkInterfaces []*armnetwork.Interface, subnets ipamTypes.SubnetMap) *ipamTypes.Instance {
```
Is there a specific reason to have an unexported implementation of `ParseInterfacesIntoInstance`? This wrapper does not add any logic.

Same for `ParseInterfacesIntoInstanceMap` and `ListAllNetworkInterfaces`.

yuecong replied:
good catch. let me make the refactoring.

pippolo84 commented on pkg/azure/ipam/instances.go:
```diff
@@ -86,67 +93,126 @@ func (m *InstancesManager) Resync(ctx context.Context) time.Time {
+	for subnetID := range subnetIDs {
+		result = append(result, subnetID)
+	}
+	return result
```
Nit: this can be simplified with:

```go
	return slices.Collect(maps.Keys(subnetIDs))
```

pippolo84 reviewed (commented):
Overall LGTM, thanks!

Left a couple of nits and a question inline.

pippolo84 reviewed (approved):
💯
